x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/model/rule_assets/prebuilt_rule_asset.ts

```diff
@@ -37,28 +30,26 @@ const BASE_PROPS_REMOVED_FROM_PREBUILT_RULE_ASSET = zodMaskFor<BaseCreateProps>(
   'outcome',
 ]);
 
-// `response_actions` is only part of the optional fields in QueryRuleCreateFields and SavedQueryRuleCreateFields
-const TYPE_SPECIFIC_PROPS_REMOVED_FROM_PREBUILT_RULE_ASSET = zodMaskFor<
-  QueryRuleCreateFields | SavedQueryRuleCreateFields
->()(['response_actions']);
-
-const QueryRuleAssetFields = QueryRuleCreateFields.omit(
-  TYPE_SPECIFIC_PROPS_REMOVED_FROM_PREBUILT_RULE_ASSET
-);
-const SavedQueryRuleAssetFields = SavedQueryRuleCreateFields.omit(
-  TYPE_SPECIFIC_PROPS_REMOVED_FROM_PREBUILT_RULE_ASSET
-);
-
-export const RuleAssetTypeSpecificCreateProps = z.discriminatedUnion('type', [
-  EqlRuleCreateFields,
-  QueryRuleAssetFields,
-  SavedQueryRuleAssetFields,
-  ThresholdRuleCreateFields,
-  ThreatMatchRuleCreateFields,
-  MachineLearningRuleCreateFields,
-  NewTermsRuleCreateFields,
-  EsqlRuleCreateFields,
-]);
+/**
+ * Aditionally remove fields which are part only of the optional fields in the rule types that make up
+ * the TypeSpecificCreateProps discriminatedUnion, by using a Zod transformation which extracts out the
+ * necessary fields in the rules types where they exist. Fields to extract:
+ *  - response_actions: from Query and SavedQuery rules
+ */
+const TypeSpecificFields = TypeSpecificCreateProps.transform((val) => {
+  switch (val.type) {
+    case 'query': {
+      const { response_actions: _, ...rest } = val;
+      return rest;
+    }
+    case 'saved_query': {
+      const { response_actions: _, ...rest } = val;
+      return rest;
+    }
+    default:
+      return val;
+  }
+});
```

Comment:
Very cool, this is much simpler than reconstructing type specific props from scratch. And it even creates a correct type for `TypeSpecificFields` which I didn't expect! Also it's extensible: this way we could omit any other fields from other rule types. Very-very cool!